### PR TITLE
log rawTx when fail to send tx

### PR DESCRIPTION
**Pull request by dai1975**
log rawTx when fail to send tx

### CODE REVIEW

File: pkg/relay/ethereum/log.go [hunk 1]
```diff
@@ -2,7 +2,7 @@ package ethereum
 
 const (
 	logAttrMsgIndexFrom = "msg_index_from"
-	logAttrMsgIndexTo   = "msg_index_to"
```
Comment (dai1975) [inline pkg/relay/ethereum/log.go:5]: change info for readers to avoid from confusing to is inclusive or not.

File: pkg/relay/ethereum/tx.go [hunk 2]
```diff
@@ -406,7 +413,7 @@ func (c *Chain) getRevertReasonFromReceipt(ctx context.Context, receipt *client.
 
 func (c *Chain) getRevertReasonFromRpcError(err error) (string, []byte, error) {
 	if de, ok := err.(rpc.DataError); !ok {
-		return "", nil, fmt.Errorf("failed with unexpected error type: errorType=%T", err)
```
Comment (dai1975) [inline pkg/relay/ethereum/tx.go:409]: output normal log in case of failing to rpc call itself.

File: pkg/relay/ethereum/tx.go [hunk 3]
```diff
@@ -504,107 +518,105 @@ func (iter *CallIter) Current() sdk.Msg {
...
 	return len(iter.msgs) <= iter.cursor
 }
-func (iter *CallIter) Next() {
-	if !iter.End() {
-		iter.cursor += 1
+func (iter *CallIter) Next(n int) {
+	iter.cursor = min(len(iter.msgs), iter.cursor + n)
+}
+
+func (iter *CallIter) updateLoggerMessageInfo(logger *log.RelayLogger, from int, count int) {
```
Comment (dai1975) [inline pkg/relay/ethereum/tx.go:525]: Being buildTx more common function, caller of buildTx need to add info of msgs in using buildTx to logger.
This is helper function for it.

File: pkg/relay/ethereum/tx.go [hunk 4]
```diff
@@ -504,107 +518,105 @@ func (iter *CallIter) Current() sdk.Msg {
...
 	// gas estimation
 	{
 		opts.GasLimit = math.MaxUint64
 		opts.NoSend = true
-		tx, err := c.SendTx(opts, iter.Current(), iter.skipUpdateClientCommitment)
+		tx, err := c.BuildMessageTx(opts, iter.Current(), iter.skipUpdateClientCommitment)
 		if err != nil {
 			logger.Error("failed to build tx for gas estimation", err)
-			return nil, err
+			return nil, 1, err
```
Comment (siburu) [inline pkg/relay/ethereum/tx.go:566]: The second return value of `buildSingleTx` sometimes becomes `0` and at other times takes on different values.
I couldn't understand the criteria for that.

File: pkg/relay/ethereum/tx.go [hunk 5]
```diff
@@ -504,107 +518,105 @@ func (iter *CallIter) Current() sdk.Msg {
...
 	// now iter.cursor < len(iter.msgs)
 
 	logger := c.GetChainLogger()
 
 	opts, err := c.TxOpts(ctx, true);
 	if err != nil {
-		if err != nil {
-			return nil, err
-		}
+		return nil, 1, err
```
Comment (siburu) [inline pkg/relay/ethereum/tx.go:595]: The second return value of `buildMultiTx` sometimes becomes `0` and at other times takes on different values.
I couldn't understand the criteria for that.

File: pkg/relay/ethereum/tx.go [hunk 6]
```diff
@@ -504,107 +518,105 @@ func (iter *CallIter) Current() sdk.Msg {
...
 	}
```
Comment (siburu) [inline pkg/relay/ethereum/tx.go:596]: This function never sends tx to the network, so you can do `opts.NoSend = true` just after calling `c.TxOpts` and checking that it doesn't return an error.

**Review by siburu — COMMENTED** [review]
@dai1975 Thank you for the PR. I left some comments.

File: pkg/relay/ethereum/tx.go [hunk 7]
```diff
@@ -504,107 +518,105 @@ func (iter *CallIter) Current() sdk.Msg {
...
+func (iter *CallIter) updateLoggerMessageInfo(logger *log.RelayLogger, from int, count int) {
+	if from < 0 || count < 0 || len(iter.msgs) <= from || len(iter.msgs) < from + count {
+		logger.Error("invalid parameter", fmt.Errorf("out of index: len(msgs)=%d, from=%d, count=%d", len(iter.msgs), from, count))
+		return
 	}
+
+	logger.Logger = logger.With(
+		logAttrMsgIndexFrom, from,
+		logAttrMsgCount, count,
+		logAttrMsgType, strings.Join(iter.msgTypeNames[0 : from + count], ","),
```
[the remark this answers is not in the document]
Comment (dai1975) [reply pkg/relay/ethereum/tx.go]: fixed

File: pkg/relay/ethereum/tx.go (hunk 4 above)
Comment (dai1975) [reply pkg/relay/ethereum/tx.go]: Its value used to inform parent caller to add log info.
When fail to build, if the error is caused by message, return message count=1 for logging, but the error is unrelated to message itself, returns message count=0.

Rewriting this behavior that logging not at caller but place of error occuring, not to return logging info.
And the function returns count only if buildTx is succeeded.

File: pkg/relay/ethereum/tx.go (hunk 5 above)
Comment (dai1975) [reply pkg/relay/ethereum/tx.go]: same as above.

File: pkg/relay/ethereum/tx.go (hunk 6 above)
Comment (dai1975) [reply pkg/relay/ethereum/tx.go:600]: modified.

**Review by siburu — APPROVED** [review]
@dai1975 LGTM. Thank you!

**Pull request merged**